tuxemon/tools.py

```diff
@@ -382,7 +382,7 @@ def copy_dict_with_keys(
     return {k: source[k] for k in keys if k in source}
 
 
-def assert_never(value: NoReturn) -> NoReturn:
+def assert_never(value: str) -> NoReturn:
```

Comment:
This change is wrong, there is no warning anymore.

Reply:
https://peps.python.org/pep-0484/#the-noreturn-type

NoReturn shouldn't be used for arguments. I will doublecheck what this function is doing, but it should probably be Any.

Reply:
That function is used to detect that every possible value of that variable has been handled at that point. That is why the type is `NoReturn` (which really means "no type allowed").

Reply:
I get this when using mypy 0.761

tuxemon/states/combat/combat.py:353: error: Argument 1 to "assert_never" has incompatible type "Union[Literal['begin'], Literal['ready'], Literal['housekeeping phase'], Literal['decision phase'], Literal['pre action phase'], Literal['action phase'], Literal['post action phase'], Literal['resolve match'], Literal['ran away'], Literal['draw match'], Literal['has winner'], Literal['end combat']]"; expected "NoReturn"

The NoReturn as a return type makes sense, but you can't use NoReturn as a typehint for "value".

See https://peps.python.org/pep-0484/#the-noreturn-type

> The NoReturn type is only valid as a return annotation of functions, and considered an error if it appears in other positions:

```
from typing import List, NoReturn

# All of the following are errors
def bad1(x: NoReturn) -> int:
    ...
bad2 = None  # type: NoReturn
def bad3() -> List[NoReturn]:
```

The signature should be `def assert_never(value: Any) -> NoReturn:` as you want to be able to use assert_never on any type.

Reply:
Then it means that these literals have not been handled. If you change the type, the function is useless.

There is a similar function in Python 3.11, but we cannot use it yet.

Reply:
```
from typing import Any, Literal, NoReturn

CombatPhase = Literal[
    "begin",
]


def assert_never(value: NoReturn) -> NoReturn:
    assert False


def outer(phase: CombatPhase):
    if phase == "begin":
        return
    else:
        assert_never(phase)
```

example.py:16: error: Argument 1 to "assert_never" has incompatible type "Literal['begin']"; expected "NoReturn"

Reply:
Oh, you're right.

https://typing.readthedocs.io/en/latest/source/unreachable.html